casper/src/main/scala/coop/rchain/casper/Validate.scala

```diff
@@ -243,33 +252,36 @@ object Validate {
     */
   def repeatDeploy[F[_]: Monad: Log: BlockStore](
       block: BlockMessage,
-      dag: BlockDagRepresentation[F]
+      dag: BlockDagRepresentation[F],
+      expirationThreshold: Int
   ): F[Either[InvalidBlock, ValidBlock]] = {
     val deployKeySet = (for {
       bd <- block.body.toList
       r  <- bd.deploys.flatMap(_.deploy)
     } yield (r.user, r.timestamp)).toSet
 
     for {
-      initParents <- ProtoUtil.unsafeGetParents[F](block)
+      initParents         <- ProtoUtil.unsafeGetParents[F](block)
+      maxBlockNumber      = ProtoUtil.maxBlockNumber(initParents)
+      earliestBlockNumber = maxBlockNumber + 1 - expirationThreshold
       duplicatedBlock <- DagOperations
-                          .bfTraverseF[F, BlockMessage](initParents)(ProtoUtil.unsafeGetParents[F])
-                          .find(
-                            _.body.exists(
-                              _.deploys
-                                .flatMap(_.deploy)
-                                .exists(
-                                  p => deployKeySet.contains((p.user, p.timestamp))
-                                )
-                            )
+                          .bfTraverseF[F, BlockMessage](initParents)(
+                            b =>
+                              ProtoUtil.unsafeGetParentsAboveBlockNumber[F](b, earliestBlockNumber)
                           )
+                          .find { b =>
+                            ProtoUtil
+                              .deploys(b)
+                              .flatMap(_.deploy)
+                              .exists(d => deployKeySet.contains((d.user, d.timestamp)))
+                          }
       result <- duplicatedBlock match {
                  case Some(b) =>
                    for {
                      _ <- Log[F].warn(
                            ignore(
                              block,
-                             s"found deploy by the same (user, millisecond timestamp) produced in the block(${b.blockHash})"
+                             s"found deploy by the same (user, millisecond timestamp) produced in the block ${b.blockHash}"
```

Comment:
Instead of this `for-comprehension` you can do this:
```scala
                   Log[F].warn(
                           ignore(
                             block,
                             s"found deploy by the same (user, millisecond timestamp) produced in the block ${b.blockHash}"
                           )
                         ) *> Left(InvalidRepeatDeploy).pure
```

The operator *> produces the effects from its both sides and returns the value on its right side.

Even more, where possible, I try to avoid the matches because they are error prone. For this reason I would replace the whole match like this:

```scala
      result <- duplicatedBlock
                 .traverse( b =>
                     Log[F].warn(
                       ignore(
                         block,
                         s"found deploy by the same (user, millisecond timestamp) produced in the block ${b.blockHash}"
                       )
                     )
                     *> InvalidRepeatDeploy.pure[F]
                 )
    } yield result.toLeft(Valid)
```